contracts/staking/KeepStake.sol

```diff
@@ -3,17 +3,83 @@
+///         but does not expose it externally.
+///
+///         KeepStake contract addresses this problem by exposing
+///         operator-owner mappings snapshotted off-chain based on events and
+///         information publicly available from KEEP TokenStaking contract and
+///         KEEP TokenGrant contract. Additionally, it gives the Governance
+///         ability to add new mappings in case they are ever needed; in
+///         practice, this will be needed only if someone decides to stake their
+///         KEEP token grant in KEEP network after 2021-11-11 when the snapshot
+///         was taken.
+///
+///         Operator-owner pairs were snapshotted 2021-11-11 in the following
+///         way:
+///         1. Fetch all TokenStaking events from KEEP staking contract.
+///         2. Filter out undelegated operators.
+///         3. Filter out canceled delegations.
+///         4. Fetch grant stake information from KEEP TokenGrant for that
+///            operator to determine if we are dealing with grant delegation.
+///         5. Fetch grantee address from KEEP TokenGrant contract.
+///         6. Check if we are dealing with ManagedGrant by looking for all
+///            created ManagedGrants and comparing their address against grantee
+///            address fetched from TokenGrant contract.
+contract KeepStake is Ownable {
+    IKeepTokenStaking public immutable keepTokenStaking;
+
+    mapping(address => address) public operatorToManagedGrant;
+    mapping(address => address) public operatorToGrantee;
+
+    constructor(IKeepTokenStaking _keepTokenStaking) {
+        keepTokenStaking = _keepTokenStaking;
+    }
+
+    /// @notice Allows the Governance to set new operator-managed grant pair.
+    ///         This function should only be called for managed grants if
+    ///         the snapshot does include this pair.
+    function setManagedGrant(address operator, address managedGrant)
         external
-        view
-        returns (address)
+        onlyOwner
     {
+        operatorToManagedGrant[operator] = managedGrant;
+    }
+
+    /// @notice Allows the Governance to set new operator-grantee pair.
+    ///         This function should only be called for non-managed grants if
+    ///         the snapshot does include this pair.
+    function setGrantee(address operator, address grantee) external onlyOwner {
+        operatorToGrantee[operator] = grantee;
+    }
+
+    /// @notice Resolves KEEP stake owner for the provided operator address.
+    function resolveOwner(address operator) external view returns (address) {
+        address owner = operatorToManagedGrant[operator];
+        if (owner != address(0)) {
+            return IManagedGrant(owner).grantee();
+        }
+        owner = operatorToGrantee[operator];
+        if (owner != address(0)) {
+            return owner;
+        }
+
         if (operator == 0x855A951162B1B93D70724484d5bdc9D00B56236B) {
```

Comment:
What do you think about add these operators to mappings in the constructor? This will be free from gas and contact size perspective and will make `resolveOwner` simpler

Reply:
Hmm... I've left them here exactly for this reason. If we add them in the constructor, the deployment will be expensive, because of SSTOREs, right? Right now, it's ~2k to resolve the owner from what I checked.

Reply:
These values are stored in contract size, so it's also cost something, but not sure more or less than SSTORE.
Also I think resolving will read less, because will be only two reads from mapping instead of two reads + ifs (in case of some static value). Of course it's only guessing, need to measure to be sure. My main point that will be easier to review this method with simpler logic.
But this is not critical

Reply:
I tried experimenting with gas costs but Remix refuses to deploy my contracts... I extracted snapshotted addresses to separate functions in https://github.com/threshold-network/solidity-contracts/pull/27/commits/6ada9c9b193bcb139a71f840c59fd8080bb8f1c5. They are internal so should be copied by the compiler anyway but at least reading the code should be easier now. We can revisit and compare costs once #22 is merged - hardhat deploy plugin displays the gas cost of this operation.